### PR TITLE
[SPARK-12295] [SQL] external spilling for window functions

### PR DESCRIPTION
**Pull request by davies**
This PR manage the memory used by window functions (buffered rows), also enable external spilling.

After this PR, we can run window functions on a partition with hundreds of millions of rows with only 1G.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 1]
```diff
@@ -380,8 +408,36 @@ private[execution] final case class RangeBoundOrdering(
...
+      inputIndex: Int,
+      outputRow: InternalRow,
+      outputIndex: Int): Int =
+    ordering.compare(current(inputRow), bound(outputRow))
+}
+
+/**
+  * An external buffer of rows based on UnsafeExternalSorter
+  */
+class ExternalRowBuffer(sorter: UnsafeExternalSorter, numFields: Int) {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: private[execution]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 2]
```diff
@@ -661,29 +719,35 @@ private[execution] final class UnboundedFollowingWindowFunctionFrame(
...
+
+    // Duplicate the input to have a new iterator
+    val tmp = input.copy()
 
     // Drop all rows from the buffer for which the input row value is smaller than
     // the output row lower bound.
-    while (inputIndex < input.size && lbound.compare(input, inputIndex, outputIndex) < 0) {
+    var i = 0
+    while (i < inputIndex) {
+      tmp.next()
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: We are skipping all the records we have already processed here; this shouldn't be to expensive; but it still adds some complexity (```O(n * (n - 1) /2)``` towards ```O(n * n)```). I'd like to see/do a few benchmarks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 3]
```diff
@@ -498,7 +548,13 @@ private[execution] final class SlidingWindowFunctionFrame(
...
 
   /** Rows of the partition currently being processed. */
-  private[this] var input: ArrayBuffer[InternalRow] = null
+  private[this] var input: ExternalRowBuffer = null
+
+  /** The next row from `input`. */
+  private[this] var nextRow: InternalRow = null
+
+  /** The rows within current sliding window. */
+  private[this] val buffer: util.Queue[InternalRow] = new util.ArrayDeque[InternalRow]()
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Nit/Performance Paranoia: I would not type it as a ```Queue``` but as an ```ArrayDeque``` (or have scala infer the type). In order to ever prevent megamorphic calls.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala (hunk 2 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Yes, it doubles the time. If N is large, it's kind of like double of infinity is still infinity :-) 

Actually I tried to add a `skip(n)` API for UnsafeSortIterator to improve this a little bit, but it seems not worth that complexity, so removed.

I think the next step could be have a fast path for those functions that has communitativity, we could reduce the complexity to `O(N * 2)`

**Comment by hvanhovell** [comment]
@davies this is pretty awesome! I have taken long look at the window code and it looks solid. I am less of an expert on the Memory management front, so maybe someone else should take a look at that.

I do have one small concern: I am absolutely convinced that in the case of large partition sizes this will outperform the current implementation by a margin. However, I am wondering what happens if we consider smaller partition sizes (e.g. n 2-32). We might take a small hit in these cases, because of the added complexity. Have you done some benchmarking on this? If you haven't this is a link to benchmark I used for my initial window prototype: https://issues.apache.org/jira/secure/attachment/12745984/perf_test3.scala

I'd like to finish with something we should **not** address in this PR (thinking out loud if you will). The child node of a ```Window``` operator is allmost always an ```ExternalSort``` operator. Wouldn't it be cool if we could eliminate the row buffer of the ```Window``` by using the external sorts buffer?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala (hunk 2 above)
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: true that :)

This is fine for now; it is not used that much.

**Comment by davies** [comment]
@hvanhovell Here is the result after having a fast path for small partition:

```
+-----+----------------+------+-----------+-----+
| Size|            Name|Master|SPARK-12295| Diff|
+-----+----------------+------+-----------+-----+
|    2|Entire Partition|  2007|       1933|0.963|
|    2|         Growing|  2576|       2193|0.851|
|    2|       Shrinking|  2945|       2829|0.961|
|    2|         Sliding|  2206|       2320|1.052|
|    8|Entire Partition|  2097|       1913|0.912|
|    8|         Growing|  1959|       2235|1.141|
|    8|       Shrinking|  2099|       2359|1.124|
|    8|         Sliding|  2124|       2040|0.960|
|   32|Entire Partition|  2261|       2260|1.000|
|   32|         Growing|  2143|       2218|1.035|
|   32|       Shrinking|  2357|       2401|1.019|
|   32|         Sliding|  2691|       2370|0.881|
|  128|Entire Partition|  2146|       1907|0.889|
|  128|         Growing|  2019|       1928|0.955|
|  128|       Shrinking|  3085|       3725|1.207|
|  128|         Sliding|  2040|       2110|1.034|
|  256|Entire Partition|  1962|       2042|1.041|
|  256|         Growing|  2777|       2335|0.841|
|  256|       Shrinking|  4179|       5027|1.203|
|  256|         Sliding|  2106|       2155|1.023|
| 1024|Entire Partition|  1710|       1734|1.014|
| 1024|         Growing|  1873|       2022|1.080|
| 1024|       Shrinking| 10456|      14588|1.395|
| 1024|         Sliding|  1928|       1917|0.994|
| 4096|Entire Partition|  1769|       1670|0.944|
| 4096|         Growing|  1986|       1844|0.928|
| 4096|       Shrinking| 39184|      67151|1.714|
| 4096|         Sliding|  2080|       2108|1.013|
|16192|Entire Partition|  2835|       1805|0.637|
|16192|         Growing|  2189|       1851|0.846|
|16192|       Shrinking|151624|     337517|2.226|
|16192|         Sliding|  4734|       5615|1.186|
+-----+----------------+------+-----------+-----+
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 4]
```diff
@@ -307,27 +314,63 @@ case class Window(
...
+          } else {
+            rows.clear()
+          }
+
           while (nextRowAvailable && nextGroup == currentGroup) {
-            rows += nextRow.copy()
+            if (sorter == null) {
+              rows += nextRow.copy()
+
+              if (rows.length >= 4096) {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Should we make this configurable?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 5]
```diff
@@ -283,23 +287,26 @@ case class Window(
...
-            nextGroup = EmptyRow
+            nextRow = null
+            nextGroup = null
           }
         }
         fetchNextRow()
 
         // Manage the current partition.
-        val rows = ArrayBuffer.empty[InternalRow]
+        val rows = ArrayBuffer.empty[UnsafeRow]
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Minor (tiny): We could set the initial size to the value at which we start using the external sorter (4096). Or we could use an array for that matter.

**Comment by hvanhovell** [comment]
@davies I left some final comments; feel free to use/ignore them.

The benchmark looks good. The shrinking case was expected to perform horrible, and it does. The other figures are so close to the old figure that I don't think we can safely say that they are different (given the limited scope and duration of the test). I was wondering, how does this perform without the optimized (hybrid) approach?

LGTM (pending review of the ```ExternalSort*``` modifications)

**Comment by davies** [comment]
@hvanhovell Before the fast path, tiny partitions look horrible:

```
+-----+----------------+------+----------+-----+
| Size|            Name|Master|This PR(before fast path) | Diff|
+-----+----------------+------+----------+-----+
|    2|Entire Partition|  2007|     12395|6.176|
|    2|         Growing|  2576|     15729|6.106|
|    2|       Shrinking|  2945|     13376|4.542|
|    2|         Sliding|  2206|     13419|6.083|
|    8|Entire Partition|  2097|      4869|2.322|
|    8|         Growing|  1959|      4495|2.295|
|    8|       Shrinking|  2099|      5021|2.392|
|    8|         Sliding|  2124|      4739|2.231|
|   32|Entire Partition|  2261|      2750|1.216|
|   32|         Growing|  2143|      3062|1.429|
|   32|       Shrinking|  2357|      3233|1.372|
|   32|         Sliding|  2691|      3270|1.215|
|  128|Entire Partition|  2146|      1964|0.915|
|  128|         Growing|  2019|      2331|1.155|
|  128|       Shrinking|  3085|      4845|1.571|
|  128|         Sliding|  2040|      2406|1.179|
|  256|Entire Partition|  1962|      2110|1.075|
|  256|         Growing|  2777|      2479|0.893|
|  256|       Shrinking|  4179|      7733|1.850|
|  256|         Sliding|  2106|      2344|1.113|
| 1024|Entire Partition|  1710|      2367|1.384|
| 1024|         Growing|  1873|      2096|1.119|
| 1024|       Shrinking| 10456|     22028|2.107|
| 1024|         Sliding|  1928|      2456|1.274|
| 4096|Entire Partition|  1769|      2062|1.166|
| 4096|         Growing|  1986|      1941|0.977|
| 4096|       Shrinking| 39184|     75981|1.939|
| 4096|         Sliding|  2080|      2376|1.142|
|16192|Entire Partition|  2835|      1630|0.575|
|16192|         Growing|  2189|      1801|0.823|
|16192|       Shrinking|151624|    305837|2.017|
|16192|         Sliding|  4734|      4430|0.936|
+-----+----------------+------+----------+-----+
```

**Comment by davies** [comment]
@JoshRosen Could you help to review the ExternalSort* parts?

**Comment by davies** [comment]
I'm going to merge this to avoid conflict, any comments will be addressed by follow-up PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 6]
```diff
@@ -307,27 +314,63 @@ case class Window(
...
+            rowBuffer = new ExternalRowBuffer(sorter, inputFields)
+          } else {
+            rowBuffer = new ArrayRowBuffer(rows)
+          }
 
           // Setup the frames.
           var i = 0
           while (i < numFrames) {
-            frames(i).prepare(rows)
+            frames(i).prepare(rowBuffer.copy())
```
Comment (tedyu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: For ExternalRowBuffer, can this copy() be saved ?

Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: @tedyu both buffers implementations are statefull (they contain an iterator). Each frame traverses the partition in a different way, so we cannot share iterators between them.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala (hunk 4 above)
Comment (tedyu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: +1 on making this configurable

**Pull request closed without merging**